### PR TITLE
adapter,storage: add `envelope` column to `mz_sources`

### PR DESCRIPTION
**Pull request by aljoscha**
### Motivation

Known-desired feature: https://github.com/MaterializeInc/cloud/issues/4245.

### Tips for reviewer

We might want to think about what strings we use for the envelopes. @ggnall may or may not have opinions about that.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by aljoscha** [comment]
Joe, I added you because of potential migration/upgrade concerns.

**Comment by aljoscha** [comment]
> @aljoscha Do you mind if I push a test directly to this branch?

@jkosh44 please go ahead!

**Review by benesch — COMMENTED** [review]
We should think ahead to what this looks like when the envelopes grow new options. `ENVELOPE DEBEZIUM` already has some unsafe options about the transaction metadata topic. `ENVELOPE UPSERT` may learn to allow you to specify the key to order by. How are we going to represent that information? A new `mz_envelopes` table that gets foreign keyed in here under a new `envelope_id` column? Additional nullable columns for each envelope type's fields? Similar questions apply for formats, which I imagine we'll want to expose in the system catalog eventually too.

**Comment by jkosh44** [comment]
It looks like the builtin migration tests are now failing on this branch. I'm looking into what exactly went wrong.

**Comment by jkosh44** [comment]
> It looks like the builtin migration tests are now failing on this branch. I'm looking into what exactly went wrong.

Looks like there was a bug in how we generate the topological sort. I have a fix locally that I'll push soon.

File: src/adapter/src/catalog.rs [hunk 1]
```diff
@@ -1497,6 +1497,49 @@ impl Source {
...
+            // NOTE(aljoscha): We could move the block for ingestsions into
+            // `SourceEnvelope` itself, but that one feels more like an internal
+            // thing and adapter should own how we represent envelopes as a
+            // string? It would not be hard to convince me otherwise, though.
+            DataSourceDesc::Ingestion(ingestion) => match ingestion.desc.envelope() {
+                SourceEnvelope::None(_) => {
+                    // WIP: Should this be "none"? This envelope is the
+                    // default but users can explicitly specify it using
+                    // `ENVELOPE NONE`.
+                    Some("append-only")
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply src/adapter/src/catalog.rs:1516]: I was thinking about the same `NULL`/`none` problem but will change to `none`. Consistency is nice here!

**Comment by aljoscha** [comment]
> We should think ahead to what this looks like when the envelopes grow new options. `ENVELOPE DEBEZIUM` already has some unsafe options about the transaction metadata topic. `ENVELOPE UPSERT` may learn to allow you to specify the key to order by. How are we going to represent that information? A new `mz_envelopes` table that gets foreign keyed in here under a new `envelope_id` column? Additional nullable columns for each envelope type's fields? Similar questions apply for formats, which I imagine we'll want to expose in the system catalog eventually too.

@benesch I thought we'd have a new `mz_envelopes`, but with `source_id` as the foreign key. And (🌶️ take): we could use _one_ JSON column for the per-envelope options. This can be problematic in its own ways, but gets around the problem of having the union of all possible envelope fields as column fields where most of them will be `NULL`.

**Comment by jkosh44** [comment]
> > It looks like the builtin migration tests are now failing on this branch. I'm looking into what exactly went wrong.
> 
> Looks like there was a bug in how we generate the topological sort. I have a fix locally that I'll push soon.

Once https://github.com/MaterializeInc/materialize/pull/16605 merges, then this should start passing CI once you rebase/merge.

**Comment by aljoscha** [comment]
> > > It looks like the builtin migration tests are now failing on this branch. I'm looking into what exactly went wrong.
> > 
> > 
> > Looks like there was a bug in how we generate the topological sort. I have a fix locally that I'll push soon.
> 
> Once #16605 merges, then this should start passing CI once you rebase/merge.

many thanks!

**Comment by benesch** [comment]
> @benesch I thought we'd have a new `mz_envelopes`, but with `source_id` as the foreign key. And (🌶️ take): we could use _one_ JSON column for the per-envelope options. This can be problematic in its own ways, but gets around the problem of having the union of all possible envelope fields as column fields where most of them will be `NULL`.

I don't mind packing a number of fields into one column, but `jsonb` is fraught, as it has a much less rich type system than the rest of SQL.  I _really_ want https://github.com/MaterializeInc/materialize/issues/6842.

Anyway, I like the idea of having `mz_source_envelopes` with a `source_id` column... or `mz_source_debezium_envelopes` with a `source_id` column and some additional metadata columns about Debezium envelopes. In the latter world, would we rather call the column you're adding here `envelope_type`, instead of `envelope`?

**Comment by aljoscha** [comment]
> Anyway, I like the idea of having `mz_source_envelopes` with a `source_id` column... or `mz_source_debezium_envelopes` with a `source_id` column and some additional metadata columns about Debezium envelopes. In the latter world, would we rather call the column you're adding here `envelope_type`, instead of `envelope`?

Sum types _do_ sound perfect for this! For the current change, I'd say we rename the column to `envelope_type`. It might even be that we can't add the column to `mz_sources` and instead need to add a new table right now (@jkosh44 & @petrosagg, it seemed we can't add the column?)

**Comment by benesch** [comment]
> Sum types _do_ sound perfect for this! For the current change, I'd say we rename the column to `envelope_type`.

This sounds great.

> It might even be that we can't add the column to mz_sources and instead need to add a new table right now (@jkosh44 & @petrosagg, it seemed we can't add the column?)

Oh, no, I think we should be able to. It's a bug in the system table migration stuff if you can't!

**Comment by aljoscha** [comment]
> > Sum types _do_ sound perfect for this! For the current change, I'd say we rename the column to `envelope_type`.
> 
> This sounds great.
> 
> > It might even be that we can't add the column to mz_sources and instead need to add a new table right now (@jkosh44 & @petrosagg, it seemed we can't add the column?)
> 
> Oh, no, I think we should be able to. It's a bug in the system table migration stuff if you can't!

@benesch see see https://materializeinc.slack.com/archives/C02FWJ94HME/p1670957380994289

**Comment by benesch** [comment]
> @benesch see see https://materializeinc.slack.com/archives/C02FWJ94HME/p1670957380994289

I think we should be able to get away with it for now, since I doubt anyone (and we can verify this) in production has run `CREATE VIEW v AS SELECT * FROM mz_sources`.

File: doc/user/content/sql/system-catalog/mz_catalog.md [hunk 2]
```diff
@@ -312,6 +312,7 @@ Field            | Type       | Meaning
 `type`           | [`text`]   | The type of the source: `kafka`, `postgres`, `load-generator`, or `subsource`.
 `connection_id`  | [`text`]   | The ID of the connection associated with the source, if any.
 `size`           | [`text`]   | The [size](/sql/create-source/#sizing-a-source) of the source.
+`envelope_type`  | [`text`]   | The [envelope](/sql/create-source/#envelopes) of the source.
```
Comment (benesch) [inline doc/user/content/sql/system-catalog/mz_catalog.md:315]:
```suggestion
`envelope_type`  | [`text`]   | The [envelope](/sql/create-source/#envelopes) of the source: `none`, `upsert`, or `debezium`.
```

**Review by benesch — APPROVED** [review]

**Comment by aljoscha** [comment]
@benesch I'd have liked to manually check all environments to see if they use `mz_sources` in (materialized) views but I won't get around to it. I'm inclined to just merge and hope, but wanted to double-check with you.

**Comment by aljoscha** [comment]
I wrote a script that connects to all environments and runs
```sql
select * from mz_views where id LIKE '%u' AND definition LIKE '%mz_sources%'
```

and 
```sql
select * from mz_materialized_views where id LIKE '%u' AND definition LIKE '%mz_sources%'
```

To see if it's used in (materialized) views. I didn't see any usage! The only possible problem would be if people used `mz_sources` in a sink, but I feel that's unlikely.

**Comment by jkosh44** [comment]
> I wrote a script that connects to all environments and runs
> 
> ```sql
> select * from mz_views where id LIKE '%u' AND definition LIKE '%mz_sources%'
> ```
> 
> and
> 
> ```sql
> select * from mz_materialized_views where id LIKE '%u' AND definition LIKE '%mz_sources%'
> ```
> 
> To see if it's used in (materialized) views. I didn't see any usage! The only possible problem would be if people used `mz_sources` in a sink, but I feel that's unlikely.

It's probably unlikely but can you check indexes too?

**Comment by benesch** [comment]
> I'm inclined to just merge and hope, but wanted to double-check with you.

Thanks for going the extra mile after all!

**Comment by aljoscha** [comment]
> > I wrote a script that connects to all environments and runs
> > ```sql
> > select * from mz_views where id LIKE '%u' AND definition LIKE '%mz_sources%'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and
> > ```sql
> > select * from mz_materialized_views where id LIKE '%u' AND definition LIKE '%mz_sources%'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > To see if it's used in (materialized) views. I didn't see any usage! The only possible problem would be if people used `mz_sources` in a sink, but I feel that's unlikely.
> 
> It's probably unlikely but can you check indexes too?

@jkosh44 I set of another script for checking the indexes (on us and eu envs) using:
```sql
select * from mz_tables t, mz_indexes i where t.name = 'mz_sources' AND i.on_id = t.id AND i.id LIKE 'u%'
```

We seem to be good!

**Pull request merged**